dspy/teleprompt/finetune_teleprompter.py

```diff
@@ -0,0 +1,377 @@
+    
+    Returns:
+        Data as a list of dictionaries with the keys `example`, `prediction`,
+        `trace`, and optionally, `score` fields. For a given example:
+        - The `example` field corresponds to the example itself.
+        - The `example_ind` field corresponds to the index of the example in the
+            `dataset`.
+        - The `prediction` field corresponds to the prediction made by the
+            program on the example.
+        - The `trace` field corresponds to the trace generated by the program
+            on the example.
+        - The `score` field corresponds to the metric score of the example, if
+            the metric is provided. Otherwise, it is not included in the data.
+    """
+    data = []
+
+    # Use Evaluate to call the program have the responses cached
+    cname = program.__class__.__name__
+    info = _INFO_BOOTSTRAP_DATA.format(len(dataset), cname, num_threads)
+    logger.info(info)
+    evaluator = Evaluate(
+        devset=dataset, num_threads=num_threads, display_progress=True, max_errors=max_errors, provide_traceback=True
+    )
+    evaluator(program, metric=metric)
+
+    data = []
+    for example in dataset:
+        data_dict = process_example(example, 0, program, metric)
+        if data_dict is not None:
+            data.append(data_dict)
+    
+    return data
+
+
+def process_example(example: Any, example_ind: int, program: Callable, metric: Optional[Callable] = None) -> Dict[str, Any]:
+    # print("Processing example:", example_ind)
+    with dspy.context(trace=[]):
+        # print("Running program...", example_ind)
+        try:
+            prediction = program(**example.inputs())
+        except Exception as e:
+            print(f"Error processing example {example_ind}: {e}")
+            return None
+        # print("Getting trace...", example_ind)
+        trace = dspy.settings.trace
+        # print("Getting score...", example_ind)
+        score = metric(example, prediction, trace) if metric else None
+
+    data_dict = {
+        'example': example,
+        'prediction': prediction,
+        'trace': trace,
+        'example_ind': example_ind
+    }
+    if metric:
+        data_dict['score'] = score
+    
+    return data_dict
+
+# TODO: If we can ensure to pass the "round" information every time a call is
+# issued to an LM, we can make repetitive un-cached calls to the same LM without
+# modifying it's temperature. This function can be removed then.
+def bootstrap_data_for_round(
+        program: Program,
+        dataset: List[Example],
+        metric: Optional[Callable[
+            [Example, Prediction, Optional[List]], Union[bool, int, float]
+        ]] = None,
+        num_threads = 1,
+        sampling_round: int = 0,
+        sampling_temperature: Optional[float] = 0.9,
+        sampling_temperature_delta: float = 0.001,
+        max_errors: int = 0
+    ) -> Union[List[Dict], AssertionError]:
+    """ Bootstrap data for the given sampling round.
```

Comment:
I'm still super surprised about the amount of docstrings here. Who is the intended audience here? If this needs to be in the docs, users won't find it here. But I think this is for internal consumption?

Let's write self-documenting code. The code should use functions and variables and structure whose content tells us what's happening. Use comments only to document unobvious decisions. Don't write a docstring that's longer than the implementation.

Reply:
Was originally meant for user facing, removed